cluster-autoscaler/core/scaledown/unneeded/nodes.go

```diff
@@ -0,0 +1,216 @@
+	limitsFinder *resource.LimitsFinder
+	cachedList   []*apiv1.Node
+	byName       map[string]*node
+}
+
+type node struct {
+	ntbr  simulator.NodeToBeRemoved
+	since time.Time
+}
+
+type scaleDownTimeGetter interface {
+	// GetScaleDownUnneededTime returns ScaleDownUnneededTime value that should be used for a given NodeGroup.
+	GetScaleDownUnneededTime(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup) (time.Duration, error)
+	// GetScaleDownUnreadyTime returns ScaleDownUnreadyTime value that should be used for a given NodeGroup.
+	GetScaleDownUnreadyTime(context *context.AutoscalingContext, nodeGroup cloudprovider.NodeGroup) (time.Duration, error)
+}
+
+// NewNodes returns a new initialized Nodes object.
+func NewNodes(sdtg scaleDownTimeGetter, limitsFinder *resource.LimitsFinder) *Nodes {
+	return &Nodes{
+		sdtg:         sdtg,
+		limitsFinder: limitsFinder,
+	}
+}
+
+// Update stores nodes along with a time at which they were found to be
+// unneeded. Previously existing timestamps are preserved.
+func (n *Nodes) Update(nodes []simulator.NodeToBeRemoved, ts time.Time) {
+	updated := make(map[string]*node, len(nodes))
+	for _, nn := range nodes {
+		name := nn.Node.Name
+		updated[name] = &node{
+			ntbr: nn,
+		}
+		if val, found := n.byName[name]; found {
+			updated[name].since = val.since
+		} else {
+			updated[name].since = ts
+		}
+	}
+	n.byName = updated
+	n.cachedList = nil
+	if klog.V(4).Enabled() {
+		for k, v := range n.byName {
+			klog.Infof("%s is unneeded since %s duration %s", k, v.since, ts.Sub(v.since))
+		}
+	}
+}
+
+// Clear resets the internal state, dropping information about all tracked nodes.
+func (n *Nodes) Clear() {
+	n.Update(nil, time.Time{})
+}
+
+// Contains returns true iff a given node is unneeded.
+func (n *Nodes) Contains(nodeName string) bool {
+	_, found := n.byName[nodeName]
+	return found
+}
+
+// AsList returns a slice of unneeded Node objects.
+func (n *Nodes) AsList() []*apiv1.Node {
+	if n.cachedList == nil {
+		n.cachedList = make([]*apiv1.Node, 0, len(n.byName))
+		for _, v := range n.byName {
+			n.cachedList = append(n.cachedList, v.ntbr.Node)
+		}
+	}
+	return n.cachedList
+}
+
+// Drop stops tracking a specified node.
+func (n *Nodes) Drop(node string) {
+	delete(n.byName, node)
+	n.cachedList = nil
+}
+
+// RemovableAt returns all nodes that can be removed at a given time, divided
+// into empty and non-empty node lists, as well as a list of nodes that were
+// unneeded, but are not removable, annotated by reason.
+func (n *Nodes) RemovableAt(context *context.AutoscalingContext, ts time.Time, resourcesLeft resource.Limits, resourcesWithLimits []string, as scaledown.ActuationStatus) (empty, needDrain []*apiv1.Node, unremovable []*simulator.UnremovableNode) {
+	nodeGroupSize := utils.GetNodeGroupSizeMap(context.CloudProvider)
+	for nodeName, v := range n.byName {
+		klog.V(2).Infof("%s was unneeded for %s", nodeName, ts.Sub(v.since).String())
+		node := v.ntbr.Node
```

Comment:
This is really the same comment as the one about cache above. I think this will work in this particular implementation, but I also think caching nodeName and getting a fresh Node from ClusterSnapshot is inherently a safer option than caching Nodes.